java/gui/src/main/java/com/squareup/subzero/SystemShutdown.java

```diff
@@ -0,0 +1,36 @@
+package com.squareup.subzero;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+
+public class SystemShutdown {
+    /**
+     * Initiate a graceful shutdown of the machine(Linux only, NOP for others(Mac)).
+     *
+     * An attempt is made to check if the hardserver is running on the machine.
+     *  - If yes, then really shutdown
+     *  - If not, it's a NOP. We probably don't want to shutdown while testing.
```

Comment:
I like the idea of having a distinction between "real" and (unit) "testing" cases. I was wondering if this logic also means if the hardserver crashes for some reason on the "real" machine, we would not be shown a graceful shutdown option.

Reply:
The hardserver is configured by default to restart immediately on shutdown. Considering we control the number of calls we do, we always start up the machine for use(clear ram and no lingering connnections, states) and this is not a network appliance, and we always test the whole software stack before a release, I believe the crash is highly unlikely.